zeroex/orderwatch/watcher.go

```diff
@@ -184,13 +185,12 @@ func (w *Watcher) setupInMemoryOrderState(signedOrder *zeroex.SignedOrder, order
 	return nil
 }
 
-// GetEvents fetches the latest order events emitted by the OrderWatcher.
-func (w *Watcher) GetEvents() []*zeroex.OrderInfo {
-	w.cachedOrderEventsMux.Lock()
-	cachedEvents := w.cachedOrderEvents
-	w.cachedOrderEvents = []*zeroex.OrderInfo{}
-	w.cachedOrderEventsMux.Lock()
-	return cachedEvents
+// Subscribe allows one to subscribe to the order events emitted by the OrderWatcher.
+// To unsubscribe, simply call `Unsubscribe` on the returned subscription.
+// The sink channel should have ample buffer space to avoid blocking other subscribers.
+// Slow subscribers are not dropped.
+func (w *Watcher) Subscribe(sink chan<- []*zeroex.OrderInfo) event.Subscription {
```

Comment:
> I know we spoke about not using channels for interfaces but in this case, the go-ethereum rpc package forced my hand. I'm still down to attempt a refactor in the future, but this approach made sense for a first iteration.

I agree with you and I think it is okay for this iteration to use channels in the exported API. However, I wanted to point out that just because the `go-ethereum/rpc` package requires the use of channels, doesn't mean that we have to use channels in the other parts of our API. We could use callbacks in the `orderwatch` package and then convert to channels in the `rpc` package.

I'm going to create an issue for re-examining our asynchronous APIs to make things more consistent. I'm okay with punting for now.

Reply:
🙏